### DISCUSSION TITLE
Extracting convex hull from combined meshes without internal edges

### DISCUSSION THREAD

**Discussion by LukeDiederich** [pyvista/pyvista#3460] [Q&A]
Hello there, 

I create a lot of small meshes (for this example we will be using extruded triangles refered to as cylinders) which I will combine afterwards to one complete mesh. 
For this operation I am refering to Issue 181  (https://github.com/pyvista/pyvista-support/issues/181). 
The code snippet from this looks as follows: 

```
import pyvista as pv

block = pv.MultiBlock()
[block.append(pv.read('Component_%d.stl' % i)) for i in range(1, 9)]

combined = block.combine(True).extract_surface()
merged = combined.clean(tolerance=0.6)
```

My Code then looks like this: 
```
block = pv.MultiBlock()
try:
    [block.append(pv.read('zylinder/Zylinder_%d.stl' % i)) for i in range(0, 9)]
finally:
    pass
combined = block.combine(True).extract_surface()
merged1 = combined.clean(tolerance=0.01)
merged.save('cylinders.stl')
```

This operation results in the given picture (viewed with a cutting plane)
![image](https://user-images.githubusercontent.com/109736841/195657565-f690141b-93b9-4bb8-b6f1-8996496579ee.png)

Now what I want from this block is only the convex hull, so the outer surfaces without the internal edges, which result from the operation. 

The results from the merge filter are basically the same, but they take longer (especially when the mesh gets really big). 
For clarification: The operation is used in my code for well over 10.000 times to combine 10.000 extruded triangles to one mesh.

I am unsure of whether Boolean Operations would do the trick, because from my understanding they need an intersection to work and for that we would have to compute the immediate neighbor for every triangle, check for intersection and then run the operation. This would take up quite some time I guess, if that works in the first place. 
I chatted with @banesullivan over this and he was skeptical with boolean operations as well. 

Attached you can find the complete mesh, as well as a folder containing the meshes one by one, from which I constructed the whole mesh. 

A quick solution we tried was this snippet:

```
block = pv.MultiBlock()
try:
    [block.append(pv.read('zylinder/Zylinder_%d.stl' % i)) for i in range(0, 9)]
finally:
    pass
block.save('cylinders.vtm')
combined = block.combine(True).extract_surface()
read = pv.read('cylinders.vtm').combine()
```
There are 2 problems with this: 
1. There are still edges inside the surface, but they are not quite different from those seen before
![image](https://user-images.githubusercontent.com/109736841/195660963-27f3bf0a-c60b-4ae9-a3f7-c78e5e93445a.png)

2. the shape changed a bit. The best case would be that we follow the exact same contour as seen in the picture. While some changes (like a smoothing effect is ok), filling complete holes which can result in the meshes(perhaps even in the middle of such meshes) is not ok. This perhaps adds another layer of complexity to the problem

The zip-folder contains:
the single file meshes (as .stl and .vtm)
the combined mesh (as .stl and .vtm)

The result I want in the end is only the convex hull from the combined meshes without the internal edges.

If anything is left unclear please ask 

Thanks in advance for any help
Cheers 
Luke
[Cylinders.zip](https://github.com/pyvista/pyvista/files/9779030/Cylinders.zip)

**Comment by darikg** [comment]
The individual components (each a triangular prism) are constructed by extruding a triangle -- are these source triangles all coplanar, and are the extrusion vectors all parallel? In that case you could just find the 2d convex hull of the original triangles, and extrude whatever that shape is.

**Reply by LukeDiederich** [reply]
In this example they are indeed, yes. But for other meshes this might not be the case for some surfaces. 
Whilst I like your approach, it only works to an extend, perhaps I wasn´t specific enough in fomulating the problem:
The triangle prisms mostly have different lengths for the extrusion operations, which is why it is not enough to find the 2d convx hull of the surface constructed by the original triangles and extrude it, because the end of the extruded prisms is not coplanar. 
Hope this clarifies it a bit more.

But thanks for the answer anyways 👍 
Cheers
Luke
